napari/plugins/_npe2.py

```diff
@@ -325,9 +334,140 @@ def on_plugins_registered(manifests: Set[PluginManifest]):
+    from .._qt.qt_viewer import QtViewer
+    from . import menu_item_template, plugin_manager
+
+    app = get_app()
+    for plugin_name, samples in plugin_manager._sample_data.items():
+        multiprovider = len(samples) > 1
+        if multiprovider:
+            sub_menu_id = f'napari/file/samples/{plugin_name}'
+            sub_menu = [
+                (
+                    MenuId.SAMPLES,
+                    SubmenuItem(
+                        submenu=sub_menu_id, title=trans._(plugin_name)
+                    ),
+                ),
+            ]
+            app.menus.append_menu_items(sub_menu)
+        else:
+            sub_menu_id = MenuId.SAMPLES
+
+        for samp_name, samp_dict in samples.items():
+
+            def _add_sample(
+                qt_viewer: QtViewer,
+                plugin=plugin_name,
+                sample=samp_name,
+            ):
+                try:
+                    qt_viewer.viewer.open_sample(plugin, sample)
+                except MultipleReaderError as e:
+                    qt_viewer._qt_open(e.paths, stack=False, plugin=plugin)
+
+            display_name = samp_dict['display_name'].replace("&", "&&")
+            if multiprovider:
+                title = display_name
+            else:
+                title = menu_item_template.format(plugin_name, display_name)
+
+            action: Action = Action(
+                id=samp_dict['display_name'],
+                title=title,
+                menus=[{'id': sub_menu_id, 'group': MenuGroup.NAVIGATION}],
+                callback=_add_sample,
+            )
+            Q_FILE_ACTIONS.append(action)
 
 
-def _register_manifest_actions(manifest: PluginManifest) -> None:
+def _build_samples_menu(mf: PluginManifest) -> None:
+    """Builds 'Open Sample' menu for a single npe2 plugin manifest."""
+    from .._app_model import get_app
+    from .._app_model.constants import MenuGroup, MenuId
+    from .._qt._qapp_model.qactions._file import Q_FILE_ACTIONS
+    from .._qt.qt_viewer import QtViewer
+    from . import menu_item_template
+
+    # If no sample data, return
+    if not mf.contributions.sample_data:
+        return
+
+    app = get_app()
+    sample_data = mf.contributions.sample_data
+    multiprovider = len(sample_data) > 1
+    if multiprovider:
+        sub_menu_id = f'napari/file/samples/{mf.name}'
+        sub_menu = [
+            (
+                MenuId.SAMPLES,
+                SubmenuItem(submenu=sub_menu_id, title=trans._(mf.name)),
+            ),
+        ]
+        app.menus.append_menu_items(sub_menu)
+    else:
+        sub_menu_id = MenuId.SAMPLES
+
+    for sample in sample_data:
+
+        def _add_sample(
+            qt_viewer: QtViewer,
+            plugin=mf.name,
+            sample=sample.key,
+        ):
+            try:
+                qt_viewer.viewer.open_sample(plugin, sample)
+            except MultipleReaderError as e:
+                qt_viewer._qt_open(e.paths, stack=False, plugin=plugin)
```

Comment:
This will go through the opening process again before popping the reader dialog. I'm not sure this is the best way. I need to revisit this and see exactly what we should call here to minimize time spent trying to open with a plugin that may not work.

Reply:
@DragaDoncila, is this still relevant?